webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts

```diff
@@ -0,0 +1,203 @@
+     */
+    public isBusy(): boolean {
+        return this.pendingIdentifiers.size > 0;
+    }
+}
+
+/**
+ * An IntervalDataLoader is an object that can be used to batch requests for fetching objects from the server. Instead
+ * of requesting data immediately, it will periodically check if any objects need to be requested from the server.
+ *
+ * It's intended to be used for loading low priority data such as information needed in response to WebSocket messages
+ * that the user won't see immediately.
+ */
+export class IntervalDataLoader<Identifier, Result = unknown> extends DataLoader<Identifier, Result> {
+    private intervalId: number = -1;
+
+    public startIntervalIfNeeded(ms: number): void {
+        if (this.intervalId !== -1) {
+            return;
+        }
+
+        this.intervalId = window.setInterval(() => this.fetchBatchNow(), ms);
+    }
+
+    public stopInterval(): void {
+        clearInterval(this.intervalId);
+        this.intervalId = -1;
+    }
+
+    public fetchBatchNow(): void {
+        const {identifiers} = this.startBatch();
+
+        if (identifiers.length === 0) {
+            return;
+        }
+
+        this.fetchBatch(identifiers);
+    }
+
+    public isBusy(): boolean {
+        return super.isBusy() || this.intervalId !== -1;
+    }
+}
+
+/**
+ * A DelayedDataLoader is an object that can be used to batch requests for fetching objects from the server. Instead of
+ * requesting data immediately, it will wait for an amount of time and then send a request to the server for all of
+ * the data which would've been requested during that time.
+ *
+ * More specifically, when queueForLoading is first called, a timer will be started. Until that timer expires, any other
+ * calls to queueForLoading will have the provided identifiers added to the ones from the initial call. When the timer
+ * finally expires, the request will be sent to the server to fetch that data. After that, the timer will be reset and
+ * the next call to queueForLoading will start a new one.
+ *
+ * DelayedDataLoader is intended to be used for loading data for components which are unaware of each other and may appear
+ * in different places in the UI from each other which could otherwise send repeated requests for the same or similar
+ * data as one another.
+ */
+export class DelayedDataLoader<Identifier> extends DataLoader<Identifier, Promise<unknown>> {
+    private readonly wait: number = -1;
+
+    private timeoutId: number = -1;
+    private timeoutCallbacks = new Set<{
+        identifiers: Set<Identifier>;
+        resolve(): void;
+    }>();
+
+    constructor(args: {
+        fetchBatch: (identifiers: Identifier[]) => Promise<unknown>;
+        maxBatchSize: number;
+        wait: number;
+    }) {
+        super(args);
+
+        this.wait = args.wait;
+    }
+
+    public queueForLoading(identifiersToLoad: Identifier[]): void {
```

Comment:
Good idea. We're continually naming these methods better since `queue` is much better than `addIdsToLoad` or whatever it was when I first submitted this